### PR TITLE
upgrade guide for sdk 2.0

### PR DESCRIPTION
**Pull request by FrankEssenberger**
## What Has Changed?
 
Introduced the update guide for sdk 2.0.

## Manual Checks?

- [x] Text adheres to the style guide (`vale docs/`)
  - Every sentence is on its own line
  - Headings use [title capitalization](https://capitalizemytitle.com/style/AP/#) (applies also to the sidebar and title of a document)
  - You followed [naming center](https://www.sapbrandtools.com/naming-center/#/dashboard) guidelines when referring to SAP products (e.g. SAP S/4HANA)
- [x] You checked your spelling and grammar (consider using Grammarly, see `CONTRIBUTING.md`)
- [x] You formatted all changed files with prettier (`npm run prettier`)
- [x] You tested if the documentation still builds (`npm run build`)
- [x] You verified all new and changed links still work (changing the `id` or name of a file can break links)
~~- [ ] You have updated the [feature matrix](https://github.com/SAP/cloud-sdk/blob/main/docs/components/data/features.js) if you add documentation on a new feature or otherwise required.~~

### CODE REVIEW

File: docs/js/guides/upgrade-to-version-2.mdx [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+We have deprecated objects in version 1.X.Y and in version 2.0.0 they were removed.
+If you use TypesScript the replacement is simple:
+- Check the objects you use.
+- Your IDE should cross out deprecated methods.
+- The JSDoc of the method will tell you how to handle the deprecated method e.g. `@deprecated Since v1.20.0. Use 'buildAuthorizationHeaders' instead.`
+
+If you use JavaScript more manual effort is needed.
+You could either proactively investigate the [deprecated object list](#deprecatedList) and replace findings.
+Alternatively, you can also proceed with the next step and see which imports fail and check if the method is deprecated then.
```
Comment (deekshas8) [inline docs/js/guides/upgrade-to-version-2.mdx:48]:
```suggestion
Alternatively, you can also proceed with the next step and see which imports fail. Then check if the method is deprecated.
```

**Comment by FrankEssenberger** [comment]
PR for imports https://github.com/SAP/cloud-sdk/pull/585

File: docs/js/guides/upgrade-to-version-2.mdx [hunk 2]
```diff
@@ -0,0 +1,290 @@
+---
+id: upgrade-to-version-2
+title: How to upgrade to version 2.0 of the SAP Cloud SDK for JavaScript
+hide_title: false
+hide_table_of_contents: false
+sidebar_label: Upgrade to version 2.0
+description: This document will guide you through the upgrade process from version 1.X.Y to 2.0.
+---
+
+## Overview
```
[the remark this answers is not in the document]
Comment (jjtang1985) [reply docs/js/guides/upgrade-to-version-2.mdx:10]: The aggregator option of the generator was removed, should we mention that?

**Review by jjtang1985 — COMMENTED** [review]
Looks promising. I left some questions/suggestions.

File: docs/js/guides/upgrade-to-version-2.mdx [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+If it does not apply to your project, please proceed to the next one.
+
+### Update Public API Clients
+
+**Prerequisite**: Your project uses published API clients like `https://www.npmjs.com/package/@sap/cloud-sdk-vdm-business-partner-service`.
+
+Follow the following steps:
+
+- Find the used clients in your `package.json`.
+- Change the used version from 1.X.Y to 2.0.0.
```
[the remark this answers is not in the document]
Comment (FrankEssenberger) [reply docs/js/guides/upgrade-to-version-2.mdx:24]: I would prefer not promoting this kind of work around.

File: docs/js/guides/upgrade-to-version-2.mdx [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+- Find the invocations of the request builders.
+- Change the `execute()` call as described [here](#executeAPIChanges)
+
+### Deprecated Methods
+
+We have deprecated objects in version 1.X.Y and in version 2.0 they were removed.
+If you use TypesScript, the replacement is simple:
+
+- Check the objects you use.
+- Your IDE should cross out deprecated methods.
```
[the remark this answers is not in the document]
Comment (FrankEssenberger) [reply docs/js/guides/upgrade-to-version-2.mdx:46]: I am not sure about this. You updated a VDM which has its own SDK with version inside. The step mentioned in this section only applies if you have explicitly installed a sdk to use it (at least if you have some checks in place that you do not use peer dependencies of dependencies diretly without declaring them)

File: docs/js/guides/upgrade-to-version-2.mdx (hunk 2 above)
[the remark this answers is not in the document]
Comment (FrankEssenberger) [reply docs/js/guides/upgrade-to-version-2.mdx:10]: added a paragraph. Regarding the exports there is already a section mentioning the `/internal`

File: docs/js/guides/upgrade-to-version-2.mdx [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+```
+
+We also try to keep the internal API stable and guarantee this for patch version increases.
+Details on the API contract are discussed in this [ADR](https://github.com/SAP/cloud-sdk-js/blob/2.0/knowledge-base/adr/0028-public-api-extraction.md)
+
+### Serializing and Deserializing of Entities
+
+**Prerequisite**: You consume non primitive entity properties like dates, duration and times.
+
+In the past, one OData datatype was always transformed to the same datatype in TypeScript.
```
Comment (marikaner) [inline docs/js/guides/upgrade-to-version-2.mdx:90]: Should we comment this out as long as it is not yet released?

File: docs/js/guides/upgrade-to-version-2.mdx [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+For example, time related properties were transformed to `moment` instances.
+We introduced a new [middle ware concept](../features/odata/execute-odata-request#setting-custom-middleware) to make this more flexible.
+
+The default serialization for time related properties is changed from `moment` to `temporal`.
+Note that in the `beta` phase of the 2.0 version, the default remains `moment`.
+Since `moment` is a discontinued project with large size we recommend to stick with the new standard.
+Please adjust your code base to use the [temporal](https://tc39.es/proposal-temporal/docs/) date/time objects.
+If this is not possible in your case please follow the [documentation](../features/odata/execute-odata-request#setting-custom-middleware) to switch back to the `moment` serialization.
+
+### You are facing issues with the Migration
```
Comment (marikaner) [inline docs/js/guides/upgrade-to-version-2.mdx:100]:
```suggestion
### You are facing issues with the Upgrade
```

**Review by marikaner — COMMENTED** [review]
I generally like the structure and I think the appendix is a good idea.
I think we could simplify it this a bit with the following ideas:
1. Introduce a TLDR; section in/after the overview that describes the changes in a few words, but links to the longer description.
2. I would like to reduce the prosa text to a minimum. Within the longer sections, I think we could state the necessary changes first and then give some background (maybe in a subsection). My ratio is, that if i am looking at this upgrade guide, I want to know as quickly as possible how to upgrade and might not necessarily be interested in the background of the change.

File: docs/js/guides/upgrade-to-version-2.mdx (hunk 3 above)
[the remark this answers is not in the document]
Comment (FrankEssenberger) [reply docs/js/guides/upgrade-to-version-2.mdx:31]: added a sentence

File: docs/js/guides/upgrade-to-version-2.mdx [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+
+- Find the imported method in your code.
+- Find the new package where the method is located as described [here](#findNewPackage).
+- Add the new package to the `package.json`
+- Remove `@sap-cloud-sdk/core`.
+- Replace the old import with package name.
+
+Note that we reduced the number of objects in the stable public API.
+Your IDE should warn you if a module has no exported member with the given name.
+In such a case, the method is reachable under the `internal` path:
```
[the remark this answers is not in the document]
Comment (FrankEssenberger) [reply docs/js/guides/upgrade-to-version-2.mdx:76]: added a clarifiation

File: docs/js/guides/upgrade-to-version-2.mdx [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+### Deprecated Methods
+
+We have deprecated objects in version 1.X.Y and in version 2.0 they were removed.
+If you use TypesScript, the replacement is simple:
+
+- Check the objects you use.
+- Your IDE should cross out deprecated methods.
+- The JSDoc of the method will tell you how to handle the deprecated method e.g. `@deprecated Since v1.20.0. Use 'buildAuthorizationHeaders' instead.`
+
+If you use JavaScript, more manual effort is needed.
```
[the remark this answers is not in the document]
Comment (FrankEssenberger) [reply docs/js/guides/upgrade-to-version-2.mdx:56]: removed the deprecated section and included it in the replcae import section.

File: docs/js/guides/upgrade-to-version-2.mdx (hunk 5 above)
Comment (FrankEssenberger) [reply docs/js/guides/upgrade-to-version-2.mdx:90]: commented out

File: docs/js/guides/upgrade-to-version-2.mdx [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+---
+
+## Overview
+
+This document will guide you through the steps necessary to upgrade to version 2.0 of the SAP Cloud SDK.
+Depending on your project, some of the steps might not be applicable.
+We have provided prerequisites at the beginning of each section.
+If it does not apply to your project, please proceed to the next one.
+
+### EcmaScript Runtime
```
[the remark this answers is not in the document]
Comment (FrankEssenberger) [reply docs/js/guides/upgrade-to-version-2.mdx]: added a bullet point section in the beginning.

File: docs/js/guides/upgrade-to-version-2.mdx [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+If you want to check a client before we publish it, follow the step below to generate it locally.
+
+### <a name="updateGeneratedClient"/> Update Generated API Clients
+
+**Prerequisite**: Your project uses generated API clients.
+
+We assume that you have installed the generator as a local `devDependency` in your project:
+
+- Find the installed generator in your `package.json`.
+- Change the used version from 1.X.Y to 2.0.0 and execute `npm i`.
```
Comment (florian-richter) [inline docs/js/guides/upgrade-to-version-2.mdx:55]:
```suggestion
- Change the used version from 1.X.Y to 2.0.0 and execute `npm install`.
```
Easier to understand if the user is not super familiar with `npm`

File: docs/js/guides/upgrade-to-version-2.mdx [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+- Regenerate the client with the updated generator.
```
Comment (florian-richter) [inline docs/js/guides/upgrade-to-version-2.mdx:56]: Link to the guide to explain how to do this.

File: docs/js/guides/upgrade-to-version-2.mdx [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+- Find the invocations of the request builders.
```
Comment (florian-richter) [inline docs/js/guides/upgrade-to-version-2.mdx:57]: Regex would be nice again

File: docs/js/guides/upgrade-to-version-2.mdx [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+- Change the `execute()` call as described [here](#executeAPIChanges)
+
+### <a name="replaceCoreImports"/> Replace Core Imports
+
+**Prerequisite**: Your project used imports from `@sap-cloud-sdk/core`
```
Comment (florian-richter) [inline docs/js/guides/upgrade-to-version-2.mdx:62]:
```suggestion
**Prerequisite**: Your project used imports from `@sap-cloud-sdk/core`. Check your `package.json` and search globally for any import statements from this package.
```

File: docs/js/guides/upgrade-to-version-2.mdx [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+
+We split the big module `@sap-cloud-sdk/core` into multiple smaller modules.
+These imports need to be adjusted:
+
+- Find the imported method in your code.
+- Find the new package where the method is located as described [here](#findNewPackage).
+- Add the new package to the `package.json`
+- Remove `@sap-cloud-sdk/core`.
```
Comment (florian-richter) [inline docs/js/guides/upgrade-to-version-2.mdx:70]:
```suggestion
- Remove `@sap-cloud-sdk/core` from the `package.json`.
```

File: docs/js/guides/upgrade-to-version-2.mdx [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+
+//Version 2.0.0
+MyEntity.requestBuillder().getAll().execute({
+  destinationName: 'myDestination',
+  jwt: 'yourJwt',
+  useCache: true
+});
+```
+
+You also see a second important detail: The optional parameter `userJwt` is renamed to `jwt`.
```
Comment (florian-richter) [inline docs/js/guides/upgrade-to-version-2.mdx:131]: Changes like this should have a header or should be part of a list so they are easily found while scanning over the document.

**Review by florian-richter — COMMENTED** [review]
Didn't read everything but here is already some feedback 😊

File: docs/js/guides/upgrade-to-version-2.mdx (hunk 15 above)
Comment (FrankEssenberger) [reply docs/js/guides/upgrade-to-version-2.mdx:131]: good point

File: docs/js/guides/upgrade-to-version-2.mdx [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+
+### <a name="findNewPackage"/> How to find the new Packages
+
+We have split the `@sap-cloud-sdk/core` package into the following smaller packages:
+
+- `@sap-cloud-sdk/connectivity` which contains all Cloud Foundry connectivity service related methods like `getDestination()`.
+- `@sap-cloud-sdk/http-client` which contains the http client with built in connectivity with `executeHttpRequest()`.
+- `@sap-cloud-sdk/odata-common` which contains the OData functionality shared between versions 2 and 4. This module is a technical module with no public API.
+- `@sap-cloud-sdk/odata-v2` which contains the OData functionality specific to OData version 2.
+- `@sap-cloud-sdk/odata-v2` which contains the OData functionality specific to OData version 4.
```
Comment (florian-richter) [inline docs/js/guides/upgrade-to-version-2.mdx:179]:
```suggestion
- `@sap-cloud-sdk/odata-v4` which contains the OData functionality specific to OData version 4.
```

File: docs/js/guides/upgrade-to-version-2.mdx [hunk 17]
```diff
@@ -0,0 +1,307 @@
...
+  { method: 'get' }
+);
+```
+
+### <a name="findNewPackage"/> How to find the new Packages
+
+We have split the `@sap-cloud-sdk/core` package into the following smaller packages:
+
+- `@sap-cloud-sdk/connectivity` which contains all Cloud Foundry connectivity service related methods like `getDestination()`.
+- `@sap-cloud-sdk/http-client` which contains the http client with built in connectivity with `executeHttpRequest()`.
```
Comment (florian-richter) [inline docs/js/guides/upgrade-to-version-2.mdx:176]:
```suggestion
- `@sap-cloud-sdk/http-client` which contains the http client with built-in connectivity with `executeHttpRequest()`.
```

File: docs/js/guides/upgrade-to-version-2.mdx [hunk 18]
```diff
@@ -0,0 +1,307 @@
...
+- [Get Support if you face a problem in the migration process](#getSupport)
+
+We have provided prerequisites at the beginning of each section.
+If it does not apply to your project, please proceed to the next one.
+
+### <a name="updateEcmaScript"/> Update EcmaScript Runtime
+
+We changed the compilation target of our source code from `es5` to `es2019`.
+Depending on your configuration this may lead to compilation errors (TypeScript) or runtime errors in (JavaScript).
+If you run a recent node version greater 12 you should be fine from a JavaScript perspective.
```
Comment (jjtang1985) [inline docs/js/guides/upgrade-to-version-2.mdx:31]:
```suggestion
If you run a recent node version greater than 12 you should be fine from a JavaScript perspective.
```
or use `>=`

File: docs/js/guides/upgrade-to-version-2.mdx [hunk 19]
```diff
@@ -0,0 +1,307 @@
...
+hide_table_of_contents: false
+sidebar_label: Upgrade to version 2.0
+description: This document will guide you through the upgrade process from version 1.X.Y to 2.0.
+---
+
+## Overview
+
+This document will guide you through the steps necessary to upgrade to version 2.0 of the SAP Cloud SDK.
+Depending on your project, some of the steps might not be applicable.
+The ToDo list is:
```
Comment (jjtang1985) [inline docs/js/guides/upgrade-to-version-2.mdx:14]:
```suggestion
The To-Do list is:
```
or to-do

**Review by jjtang1985 — APPROVED** [review]
LGTM
I left some minor comments.

File: docs/js/guides/migrate-to-os.mdx [hunk 20]
```diff
@@ -27,8 +27,8 @@ Please use the new packages from now on.
 | `@sap/cloud-sdk-generator` | `@sap-cloud-sdk/generator` |
 | `@sap/cloud-sdk-test-util` | `@sap-cloud-sdk/test-util` |
 
-This mapping applies to the move to open source version 1.X.Y..
-Since version 2.0.0, the `core` package has been split into smaller packages.
+This mapping applies to the move to open source version 1.X.Y.
+Since version 2.0.0, the `core` package has been split into smaller packages and the `analytics` package has been deprecated.
```
Comment (florian-richter) [inline docs/js/guides/migrate-to-os.mdx:31]: Is the analytics package deprecated or removed?

File: docs/js/guides/upgrade-to-version-2.mdx (hunk 18 above)
Comment (florian-richter) [reply docs/js/guides/upgrade-to-version-2.mdx:31]:
```suggestion
If you run node version 12 or greater, no changes are necessary.
If you run an older node version, please update to a newer version.
You can find a list of breaking changes in the news section of the [node.js website](https://nodejs.org/en/blog/) (e.g. ["The Difference Between Node.js 10 and 12"](https://nodejs.org/en/blog/uncategorized/10-lts-to-12-lts/))
```

File: docs/js/guides/migrate-to-os.mdx (hunk 20 above)
Comment (FrankEssenberger) [reply docs/js/guides/migrate-to-os.mdx:31]: deprecated: was released in 2 beat. However, in version 2 we will not release it. I adjust it.

**Pull request merged**